concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -0,0 +1,804 @@
+signQuorumSignatureMessage msg privKey =
+    QuorumSignature $ Bls.sign (quorumSignatureMessageBytes msg) privKey
+
+-- |Check the signature on a 'QuorumSignatureMessage' that is signed by a single party.
+checkQuorumSignatureSingle ::
+    QuorumSignatureMessage -> BakerAggregationVerifyKey -> QuorumSignature -> Bool
+checkQuorumSignatureSingle msg pubKey =
+    Bls.verify (quorumSignatureMessageBytes msg) pubKey
+        . theQuorumSignature
+
+-- |Check the signature on a 'QuorumSignatureMessage' that is signed by multiple parties.
+checkQuorumSignature ::
+    QuorumSignatureMessage -> [BakerAggregationVerifyKey] -> QuorumSignature -> Bool
+checkQuorumSignature msg pubKeys =
+    Bls.verifyAggregate (quorumSignatureMessageBytes msg) pubKeys
+        . theQuorumSignature
+
+-- |Index of a finalizer in a finalization committee.
+newtype FinalizerIndex = FinalizerIndex {theFinalizerIndex :: Word32}
+    deriving (Eq, Ord, Show, Enum, Bounded, Serialize)
+
+-- |A set of 'FinalizerIndex'es.
+-- This is represented as a bit vector, where the bit @i@ is set iff the finalizer index @i@ is
+-- in the set.
+newtype FinalizerSet = FinalizerSet {theFinalizerSet :: Natural}
+    deriving (Eq)
+
+-- |The serialization of a 'FinalizerSet' consists of a length (Word32, big-endian), followed by
+-- that many bytes, the first of which (if any) must be non-zero. These bytes encode the bit-vector
+-- in big-endian. This enforces that the serialization of a finalizer set is unique.
+instance Serialize FinalizerSet where
+    put fs = do
+        let (byteCount, putBytes) = unroll 0 (return ()) (theFinalizerSet fs)
+        putWord32be byteCount
+        putBytes
+      where
+        unroll bc cont 0 = (bc, cont)
+        unroll bc cont n = unroll (bc + 1) (putWord8 (fromIntegral n) >> cont) (shiftR n 8)
+    get = label "FinalizerSet" $ do
+        byteCount <- getWord32be
+        FinalizerSet <$> roll1 byteCount
+      where
+        roll1 0 = return 0
+        roll1 bc = do
+            b <- getWord8
+            when (b == 0) $ fail "unexpected 0 byte"
+            roll (bc - 1) (fromIntegral b)
+        roll 0 n = return n
+        roll bc n = do
+            b <- getWord8
+            roll (bc - 1) (shiftL n 8 .|. fromIntegral b)
+
+-- |Convert a 'FinalizerSet' to a list of 'FinalizerIndex'.
+finalizerList :: FinalizerSet -> [FinalizerIndex]
+finalizerList = unroll 0 . theFinalizerSet
+  where
+    unroll _ 0 = []
+    unroll i x
+        | testBit x 0 = FinalizerIndex i : r
+        | otherwise = r
+      where
+        r = unroll (i + 1) (shiftR x 1)
+
+instance Show FinalizerSet where
+    show = show . finalizerList
+
+data QuorumCertificate = QuorumCertificate
+    { qcBlock :: !BlockHash,
+      qcRound :: !Round,
+      qcEpoch :: !Epoch,
+      qcAggregateSignature :: !QuorumSignature,
+      qcSignatories :: !FinalizerSet
+    }
+    deriving (Eq, Show)
```

Comment:
Please document, including all the fields.

Reply:
Documentation added.